File: exercises/k_closest_origin_973.py
import heapq
from typing import List

class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        heap = []
        for i, (x, y) in enumerate(points[:k]):
            dist = x*x + y*y
            heapq.heappush(heap, (-dist, i))
        for i, (x, y) in enumerate(points[k:], k):
            dist = x*x + y*y
            if -dist > heap[0][0]:
                heapq.heapreplace(heap, (-dist, i))
                
        return [points[i] for _, i in heap]

File: exercises/k_closest_origin_973.java
import java.util.PriorityQueue;
class Solution {
    public int[][] kClosest(int[][] points, int k) {
		PriorityQueue<Integer[]> maxHeap = new PriorityQueue<>((a,b) -> Integer.compare(b[0], a[0]));
		int[][] res = new int[k][2];
		for (int i = 0; i < points.length; i++) {
			int[] point = points[i];
            if (maxHeap.size() < k) {
                int value = point[0] * point[0] + (point[1] * point[1]);
                maxHeap.offer(new Integer[] {value, i});
            } else  {
                int value = point[0] * point[0] + (point[1] * point[1]);
                if (value < maxHeap.peek()[0]) {
                    maxHeap.poll();
                    maxHeap.offer(new Integer[] {value, i});
                }
            }
		}
		int i = 0;
        for (var element : maxHeap) {
            res[i] = points[element[1]];
			i++;
        }
		return res;
    }
}
